rust/rope/src/find.rs

```diff
@@ -137,81 +152,164 @@ fn find_progress_iter(cursor: &mut Cursor<RopeInfo>, pat: &str,
+            cursor.next::<BaseMetric>();
+            return None;
         }
         pat = &pat[n..];
         if pat.is_empty() {
             cursor.set(success_pos);
-            return true;
+            return Some(start_position);
         }
         let _ = cursor.next_leaf();
     }
-    false
+    cursor.set(start_position);
+    cursor.next::<BaseMetric>();
+    None
 }
 
 /// Like `compare_cursor_str` but case invariant (using to_lowercase() to
-/// normalize both strings before comparison).
-fn compare_cursor_str_casei(cursor: &mut Cursor<RopeInfo>, pat: &str) -> bool {
+/// normalize both strings before comparison). Returns the start position
+/// of the match.
+fn compare_cursor_str_casei(cursor: &mut Cursor<RopeInfo>, _lines: &mut LinesRaw, pat: &str) -> Option<usize> {
+    let start_position = cursor.pos();
     let mut pat_iter = pat.chars();
     let mut c = pat_iter.next().unwrap();
     loop {
         if let Some(rope_c) = cursor.next_codepoint() {
             for lc_c in rope_c.to_lowercase() {
                 if c != lc_c {
-                    return false;
+                    cursor.set(start_position);
+                    cursor.next::<BaseMetric>();
+                    return None;
                 }
                 if let Some(next_c) = pat_iter.next() {
                     c = next_c;
                 } else {
-                    return true;
+                    return Some(start_position);
                 }
             }
         } else {
             // end of string before pattern is complete
-            return false;
+            cursor.set(start_position);
+            cursor.next::<BaseMetric>();
+            return None;
         }
     }
 }
 
+/// Compare whether the substring beginning at the cursor location matches
+/// the provided regular expression. The substring begins at the beginning
+/// of the start of the line.
+/// If the regular expression can match multiple lines then the entire text
+/// is consumed and matched against the regular expression. Otherwise only
+/// the current line is matched. Returns the start position of the match.
+fn compare_cursor_regex(cursor: &mut Cursor<RopeInfo>, lines: &mut LinesRaw, pat: &str, cm: CaseMatching) -> Option<usize> {
+    let orig_position = cursor.pos();
+    let total_len = cursor.total_len();
+
+    if pat.is_empty() {
+        return Some(orig_position);
+    }
+
+    // create regex from untrusted input
+    let regex = RegexBuilder::new(pat)
+        .size_limit(REGEX_SIZE_LIMIT)
+        .case_insensitive(cm == CaseMatching::CaseInsensitive)
+        .build();
+
+    match regex {
+        Ok(regex) => {
+            let mut text = String::new();
+
+            if is_multiline_regex(pat) {
+                // consume all of the text if regex is muli line matching
+                text.extend(lines);
+            } else {
+                match lines.next() {
+                    Some(line) => text.push_str(&line),
```

Comment:
This seems ok for now, but is a copy of every byte of the source text. What I'd prefer is that `text` is a `Cow<str>`, and that in the single line case, it just copies the `Cow` returned by the lines iterator (meaning that if it's a borrowed reference, it just copies the reference, not the actual bytes).

Since you're reusing the buffer, at least you're not going to get an allocation per line, only a copy, which is not too bad. I wouldn't be surprised if the cost of the regex evaluation dominated in most cases.

What I think I'm most comfortable with is accepting this code as-is, and having a future work item to avoid the copy based on `Cow`.

Reply:
Okay, that sounds reasonable. I will make a separate PR for this later.